### PR TITLE
PLT-727: Update the contributing guide

### PR DESCRIPTION
**Pull request by zliu41**
Mostly the same as https://github.com/input-output-hk/plutus/pull/4867, the main difference being (1) plutus-apps doesn't allow merge commits while plutus allows all 3 merge methods; (2) branch names (master vs. main/next-node).

### CODE REVIEW

File: CONTRIBUTING.adoc [hunk 1]
```diff
@@ -307,17 +307,43 @@ Code review is great (see below), but it can slow you down if you don't take the
...
+You basically have three options:
+
+- Open the first PR against `main`, the second PR against the first PR's branch, and so on.
+  Merging a stack of PRs created this way into `main` can be non-trivial.
+- Wait until one PR is merged before opening the next PR.
+- Use a single PR for the whole feature that contains multiple small commits.
+  The problem is that Github doesn't support approving, rejecting or merging individual commits in a PR.
+  You can look at each individual commit, but it's not necessarily useful or even appropriate - many PRs have quite messy commits, and commits are sometimes overwritten via force push.
+
+The first two options are often referred to as ["trunk-based development"](https://trunkbaseddevelopment.com/), while the third "long-lived feature branches".
```
Comment (koslambrou) [inline CONTRIBUTING.adoc:328]: > The first two options are often referred to as ["trunk-based development"](https://trunkbaseddevelopment.com/)

Good to know :) I thought we mostly called "stacked diffs"

File: CONTRIBUTING.adoc [hunk 2]
```diff
@@ -336,25 +362,34 @@ It's amazing how many obvious things you spot here, and it stops the reviewer ha
...
 * Add low-effort or WIP commits to fix review comments, and then squash them away before merging the PR.
 * If you have already had a PR review, don't rebase away the old commits until the PR is ready to merge, so that the reviewer only has to look at the "new" commits.
 * Rewrite the commits to make the story clearer where possible.
-* Always prefer `git push --force-with-lease` to just `git push --force` to ensure that no work gets accidentally deleted.
-
-Don't be obsessive about history though: a little bit of effort making the history clear is nice, but you can rapidly hit diminishing returns.
-Use your judgement, but probably don't merge a PR that has commits called "WIP" or "fix"!
 
-If a PR is just a total mess, consider using Github's squash-merge feature.
+It is advisable to always prefer `git push --force-with-lease` to just `git push --force` to ensure that no work gets accidentally deleted.
```
Comment (koslambrou) [inline CONTRIBUTING.adoc:390]: to just -> instead of

File: CONTRIBUTING.adoc [hunk 3]
```diff
@@ -389,6 +424,37 @@ Read these blog posts for more good tips:
...
+
+This repo allows two merge methods: squash and merge, and rebase and merge.
+Use the one you deem appropriate.
+As said before, sometimes people use a single PR with multiple commits for their work; other times they create multiple small PRs.
+The best merge method is different for different cases.
+
+The "squash and merge" method should be used with caution.
+If you use this method, your PR must have a clean commit history: every commit should have a meaningful message, and should be buildable.
+You don't want to have commits like "fix a typo", "this may work" or "wip, done for the day" in `main` with a linear history.
+And if some of these commits are non-buildable, it can create problems for "git bisect".
```
Comment (koslambrou) [inline CONTRIBUTING.adoc:439]: Did you mix up with rebase and merge?

**Review by koslambrou — APPROVED** [review]

File: CONTRIBUTING.adoc (hunk 3 above)
Comment (zliu41) [reply CONTRIBUTING.adoc:439]: Yeah, good catch!

**Pull request merged**